enhancements/cli-in-cluster-management.md

```diff
@@ -0,0 +1,172 @@
+    * With krew we will package artifacts within an image that will populate an index.  Mirrored local registries will provide binaries to disconnected environments.  
+
+Option 4 is the option we as a team are leaning toward.  It requires that each OpenShift CLI publish artifacts, for example, [index of /pub/openshift-v4/clients](https://mirror.openshift.com/pub/openshift-v4/clients/)
+then create an image similar to these: [example images for oc cli-manager](https://github.com/sallyom/clustercli/tree/master/hack/cli-images).  CLI artifacts are packaged in images.
+The CLI in-cluster manager will gather information from each CLI's CR about each CLI and its image.  A Krew index will be populated in every cluster to provide available CLIs and plugins for a cluster.
+The logic for `oc image extract` can be re-used and extended for extracting the CLI binaries.  CLIs will be installed and managed by users through Krew, modified and extended to handle OpenShift CLIs.    
+
+Example function to be provided by Option 4 above (that will be accomplished through CLI CustomResources, CLI image, and Krew):
+
+```console
+$ oc image extract quay.io/sallyom/kn:latest --path /tmp/amd64/kn.tar.gz:$HOME/openshift-clis/compressed/
+$ tar xvf $HOME/openshift-clis/compressed/kn.tar.gz -C $HOME/openshift-clis/
+```
+
+As laid out in this proposal, each CLI will provide a separate image.
+Each CLI is responsible for creating a CR to hold metadata that `oc krew` will depend on.  The CR will serve to deliver the metadata and description
+of its deliverable binary.  An in-cluster CLI manager controller will access the CRs to populate and maintain an in-cluster Krew Index.  Users will install OpenShift tools that are known compatible
+with each cluster version through `oc krew`.    
+
+A CLI must provide a ClusterCLI CR.  The result of this proposal will be:
+* ClusterCLI Custom Resource Definition
+* Krew Index and a controller to create the index from CLI CRs.
+* mechanism to manage CLIs via ClusterCLI CRs and Krew
+* a reference consumer implementation, oc, that other CLIs can follow
+
+The CLI in-cluster management will require:
+- CLI images in an available registry- each CLI will produce an image to deliver its artifacts
+- Consumers - through an in-cluster Krew Index and the krew command, consumers will manage OpenShift CLIs
+  Consumers will require access to CLI images.  In disconnected environments, this will be a mirrored local registry. 
+- Producers - each CLI that wants to offer its binaries in-cluster is responsible for providing its metadata and description via a ClusterCLI as well as
+  a CLI image to deliver its artifacts. 
```

Comment:
In the above section I'd like to clearly spelled out:
1. CLI author requirements:
 - image containing the binaries
 - CR describing the binaries (the location, metadata, description)
2. The new operator that will be responsible for:
  - registering new CRD 
  - reading metadata from CR from authors and extracting the binaries from those images to local disk
  - serving extracted binaries
  - serving krew-compatible index
3.  oc changes
  - embed krew

Reply:
added, updated